In [19]:
# here all library that we need
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.gaussian_process.kernels import RBF
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import TfidfVectorizer
import time


In [20]:
# to make it easy to understand what im doing here i write a definition for all the algorithm 
# that i will use them in our model so instead of write that code every time i need it i just 
# need to call the def for it and give it the X and y parameter and it will work .
def destree(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    dtc = DecisionTreeClassifier()
    dtc.fit(X_train, y_train)
    y_pred =dtc.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("decision tree ",acc)
    print('Execution time:', elapsed_time, 'seconds')
    

def svc(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    svc = SVC()
    svc.fit(X_train, y_train)
    y_pred =svc.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("svc ",acc)
    print('Execution time:', elapsed_time, 'seconds')
    

def mlp(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    mlp = MLPClassifier()
    mlp.fit(X_train, y_train)
    y_pred =mlp.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("mlp",acc)
    print('Execution time:', elapsed_time, 'seconds')
    
def knb(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    knb = KNeighborsClassifier()
    knb.fit(X_train, y_train)
    y_pred =knb.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("KNeighbors",acc)
    print('Execution time:', elapsed_time, 'seconds')
    
    
def gpc(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    gpc = GaussianProcessClassifier()
    gpc.fit(X_train, y_train)
    y_pred =gpc.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("GaussianProcess",acc)
    print('Execution time:', elapsed_time, 'seconds')
    
def rbf(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    rbf = RBF()
    rbf.fit(X_train, y_train)
    y_pred =rbf.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("RBF",acc)
    print('Execution time:', elapsed_time, 'seconds')
    
def rfc(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    rfc = RandomForestClassifier()
    rfc.fit(X_train, y_train)
    y_pred =rfc.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("RandomForest",acc)
    print('Execution time:', elapsed_time, 'seconds')
    
def abc(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    abc = AdaBoostClassifier()
    abc.fit(X_train, y_train)
    y_pred =abc.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("AdaBoost",acc)
    print('Execution time:', elapsed_time, 'seconds')
    
def gnb(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    gnb = GaussianNB()
    gnb.fit(X_train, y_train)
    y_pred =gnb.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("GaussianNB",acc)
    print('Execution time:', elapsed_time, 'seconds')

def gda(X,y):
    st = time.time()
    X_train, X_test, y_train, y_test =train_test_split(X, y,test_size =0.3,random_state=42)
    gda = QuadraticDiscriminantAnalysis()
    gda.fit(X_train, y_train)
    y_pred =gda.predict(X_test)
    acc = accuracy_score(y_test ,y_pred)
    et = time.time()
    elapsed_time=et-st
    print("QuadraticDiscriminantAnalysis",acc)
    print('Execution time:', elapsed_time, 'seconds')

In [21]:
# load data by using pandas command [pd.read_csv]
data= pd.read_csv('SPAM text message.csv')
data

Category                                            Message
0         ham  Go until jurong point, crazy.. Available only ...
1         ham                      Ok lar... Joking wif u oni...
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...
3         ham  U dun say so early hor... U c already then say...
4         ham  Nah I don't think he goes to usf, he lives aro...
...       ...                                                ...
5567     spam  This is the 2nd time we have tried 2 contact u...
5568      ham               Will ü b going to esplanade fr home?
5569      ham  Pity, * was in mood for that. So...any other s...
5570      ham  The guy did some bitching but I acted like i'd...
5571      ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

## feature extraction from text 
 * Basics of CountVectorizer
 * Try one without any change in countvectorizer 
     * worst one was [ Quadratic Discriminant Analysis ] with accuracy = 77.1%
     * best one was [ MLP Classifier ] with accuracy = 98.92%

In [22]:
count_vect = CountVectorizer()
count_matrix = count_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data= count_array,columns=count_vect.get_feature_names())
df


C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8709 columns]

In [23]:
# define df as X
X=df
X

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8709 columns]

In [24]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [25]:
# [1] with desision tree 
destree(X,y)

decision tree  0.9706937799043063
Execution time: 13.149948120117188 seconds


In [26]:
# [2] with svc 
svc(X,y)

svc  0.9802631578947368
Execution time: 36.56276226043701 seconds


In [27]:
# [3] with mlp
mlp(X,y)

mlp 0.9892344497607656
Execution time: 39.728859186172485 seconds


In [28]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9156698564593302
Execution time: 1.4146463871002197 seconds


In [29]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9425837320574163
Execution time: 82.67000985145569 seconds


In [30]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [31]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9754784688995215
Execution time: 14.589043617248535 seconds


In [32]:
# [8] with abc
abc(X,y)

AdaBoost 0.9688995215311005
Execution time: 25.94694685935974 seconds


In [33]:
# [9] with gaussianNB
gnb(X,y)

GaussianNB 0.9102870813397129
Execution time: 0.6908245086669922 seconds


In [34]:
# [10] with quardratic discriminant analysis
gda(X,y)

C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


QuadraticDiscriminantAnalysis 0.7715311004784688
Execution time: 45.42710328102112 seconds


## Lower case
* Try two with change the lower case to false in countvectorizer 
     * worst one was [ Quadratic Discriminant Analysis ] with accuracy = 67.3%
     * best one was [ MLP Classifier ] with accuracy = 98.86%

In [35]:
count_vect = CountVectorizer(lowercase= False)
count_matrix = count_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data= count_array,columns=count_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zac  zealand  zed  zeros  zhong  zoom  zyada  èn  \
0              0   0  ...    0        0    0      0      0     0      0   0   
1              0   0  ...    0        0    0      0      0     0      0   0   
2              0   0  ...    0        0    0      0      0     0      0   0   
3              0   0  ...    0        0    0      0      0     0      0   0   
4              0   0  ...    0        0    0      0      0     0      0   0   
...          ...  ..  ...  ...      ...  ...    ...    ...   ...    ...  ..   
5567           0   0  ...    0        0    0      0      0     0      0   0   
5568           0   0  ...    0        0    0      0      0     0      0   0   
5569           0   0  ...    0        0    0      0      0     0      0   0   
5570           0   0  ...    0        0    0      0      0     0      0   0   
5571           0   0  ...    0        0    0      0      0     0      0   0   

      ú1  〨ud  
0      0    0  
1      0    0  
2      0    0  
3      0    0  
4      0    0  
...   ..  ...  
5567   0    0  
5568   0    0  
5569   0    0  
5570   0    0  
5571   0    0  

[5572 rows x 10857 columns]

In [36]:
# define df as X
X=df
X

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zac  zealand  zed  zeros  zhong  zoom  zyada  èn  \
0              0   0  ...    0        0    0      0      0     0      0   0   
1              0   0  ...    0        0    0      0      0     0      0   0   
2              0   0  ...    0        0    0      0      0     0      0   0   
3              0   0  ...    0        0    0      0      0     0      0   0   
4              0   0  ...    0        0    0      0      0     0      0   0   
...          ...  ..  ...  ...      ...  ...    ...    ...   ...    ...  ..   
5567           0   0  ...    0        0    0      0      0     0      0   0   
5568           0   0  ...    0        0    0      0      0     0      0   0   
5569           0   0  ...    0        0    0      0      0     0      0   0   
5570           0   0  ...    0        0    0      0      0     0      0   0   
5571           0   0  ...    0        0    0      0      0     0      0   0   

      ú1  〨ud  
0      0    0  
1      0    0  
2      0    0  
3      0    0  
4      0    0  
...   ..  ...  
5567   0    0  
5568   0    0  
5569   0    0  
5570   0    0  
5571   0    0  

[5572 rows x 10857 columns]

In [37]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [38]:
# [1] with desision tree 
destree(X,y)


decision tree  0.9736842105263158
Execution time: 18.16794204711914 seconds


In [39]:
# [2] with svc 
svc(X,y)

svc  0.9766746411483254
Execution time: 37.959561347961426 seconds


In [40]:
# [3] with mlp
mlp(X,y)

mlp 0.9886363636363636
Execution time: 37.04901933670044 seconds


In [41]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9078947368421053
Execution time: 1.3744518756866455 seconds


In [42]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9383971291866029
Execution time: 99.06560945510864 seconds


In [43]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [44]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9808612440191388
Execution time: 20.006246089935303 seconds


In [45]:
# [8] with abc
abc(X,y)

AdaBoost 0.9742822966507177
Execution time: 32.856772661209106 seconds


In [46]:
# [9] with gaussianNB
gnb(X,y)

GaussianNB 0.9360047846889952
Execution time: 0.8422331809997559 seconds


In [47]:
# [10] with quardratic discriminant analysis
gda(X,y)

C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


QuadraticDiscriminantAnalysis 0.6734449760765551
Execution time: 58.91082835197449 seconds


In [48]:
import nltk
# download punkt from nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\emana\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

## stop word we choose [is , to , my ]
* Try three with change the lower case to false in countvectorizer 
     * worst one was [ Quadratic Discriminant Analysis ] with accuracy = 75.53%
     * best one was [ MLP Classifier ] with accuracy = 98.86%

In [49]:
count_vect = CountVectorizer(stop_words=['is','to','my'])
count_matrix = count_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data= count_array,columns=count_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8706 columns]

In [50]:
# define df as X
X=df
X

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8706 columns]

In [51]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [52]:
# [1] with desision tree 
destree(X,y)

decision tree  0.9712918660287081
Execution time: 12.22007155418396 seconds


In [53]:
# [2] with svc 
svc(X,y)

svc  0.9796650717703349
Execution time: 26.395300149917603 seconds


In [54]:
# [3] with mlp
mlp(X,y)

mlp 0.9886363636363636
Execution time: 28.879722595214844 seconds


In [55]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9138755980861244
Execution time: 1.1795766353607178 seconds


In [56]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9395933014354066
Execution time: 89.50648260116577 seconds


In [57]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [58]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9766746411483254
Execution time: 14.67622971534729 seconds


In [59]:
# [8] with abc
abc(X,y)

AdaBoost 0.9736842105263158
Execution time: 25.815561532974243 seconds


In [60]:
# [9] with gaussianNB
gnb(X,y)


GaussianNB 0.9102870813397129
Execution time: 0.6655385494232178 seconds


In [61]:
# [10] with quardratic discriminant analysis
gda(X,y)

C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


QuadraticDiscriminantAnalysis 0.7553827751196173
Execution time: 44.00120973587036 seconds


## stop words sklearn built in stop words list 
* Try four with change the lower case to false in countvectorizer 
     * worst one was [ Quadratic Discriminant Analysis ] with accuracy = 48.86%
     * best one was [ MLP Classifier ] with accuracy = 98.92%

In [62]:
count_vect = CountVectorizer(stop_words='english')
count_matrix = count_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data= count_array,columns=count_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8440 columns]

In [63]:
# define df as X
X=df
X

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8440 columns]

In [64]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [65]:
# [1] with desision tree 
destree(X,y)

decision tree  0.9748803827751196
Execution time: 16.472554922103882 seconds


In [66]:
# [2] with svc 
svc(X,y)

svc  0.979066985645933
Execution time: 26.73157501220703 seconds


In [67]:
# [3] with mlp
mlp(X,y)

mlp 0.9886363636363636
Execution time: 34.26791524887085 seconds


In [68]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9084928229665071
Execution time: 1.1402599811553955 seconds


In [69]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9324162679425837
Execution time: 83.73776006698608 seconds


In [70]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [71]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9766746411483254
Execution time: 19.386964082717896 seconds


In [72]:
# [8] with abc
abc(X,y)

AdaBoost 0.9712918660287081
Execution time: 24.158165454864502 seconds


In [73]:
# [9] with gaussianNB
gnb(X,y)

GaussianNB 0.8953349282296651
Execution time: 0.64351487159729 seconds


In [74]:
# [10] with quardratic discriminant analysis
gda(X,y)

C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


QuadraticDiscriminantAnalysis 0.48863636363636365
Execution time: 42.299973249435425 seconds


## CountVectorizer using max_df 
* Try five with change the lower case to false in countvectorizer 
     * worst one was [ Quadratic Discriminant Analysis ] with accuracy = 77.15%
     * best one was [ MLP Classifier ] with accuracy = 98.92%

In [75]:
count_vect = CountVectorizer(max_df=0.5)
count_matrix = count_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data= count_array,columns=count_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8709 columns]

In [76]:
# define df as X
X=df
X

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8709 columns]

In [77]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [78]:
# [1] with desision tree 
destree(X,y)

decision tree  0.97188995215311
Execution time: 10.620443105697632 seconds


In [79]:
# [2] with svc 
svc(X,y)

svc  0.9802631578947368
Execution time: 26.18082046508789 seconds


In [80]:
# [3] with mlp
mlp(X,y)

mlp 0.9892344497607656
Execution time: 30.225085496902466 seconds


In [81]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9156698564593302
Execution time: 1.1332244873046875 seconds


In [82]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9425837320574163
Execution time: 84.9646577835083 seconds


In [83]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [84]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9772727272727273
Execution time: 14.342540264129639 seconds


In [85]:
# [8] with abc
abc(X,y)

AdaBoost 0.9688995215311005
Execution time: 24.933454751968384 seconds


In [86]:
# [9] with gaussianNB
gnb(X,y)


GaussianNB 0.9102870813397129
Execution time: 0.6696462631225586 seconds


In [87]:
# [10] with quardratic discriminant analysis
gda(X,y)

C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


QuadraticDiscriminantAnalysis 0.7715311004784688
Execution time: 43.38758444786072 seconds


## CountVectorizer using min_df 
* Try six with change the lower case to false in countvectorizer 
     * worst one was [ Quadratic Discriminant Analysis ] with accuracy = 69.79%
     * best one was [ Random forest ] with accuracy = 98.08%

In [88]:
count_vect = CountVectorizer(min_df=0.01)
count_matrix = count_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data= count_array,columns=count_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


150p  50  about  after  again  all  already  also  always  am  ...  win  \
0        0   0      0      0      0    0        0     0       0   0  ...    0   
1        0   0      0      0      0    0        0     0       0   0  ...    0   
2        0   0      0      0      0    0        0     0       0   0  ...    1   
3        0   0      0      0      0    0        1     0       0   0  ...    0   
4        0   0      0      0      0    0        0     0       0   0  ...    0   
...    ...  ..    ...    ...    ...  ...      ...   ...     ...  ..  ...  ...   
5567     0   0      0      0      0    0        0     0       0   0  ...    0   
5568     0   0      0      0      0    0        0     0       0   0  ...    0   
5569     0   0      0      0      0    0        0     0       0   0  ...    0   
5570     0   0      0      0      0    0        0     0       0   0  ...    0   
5571     0   0      0      0      0    0        0     0       0   0  ...    0   

      with  won  work  would  www  yeah  yes  you  your  
0        0    0     0      0    0     0    0    0     0  
1        0    0     0      0    0     0    0    0     0  
2        0    0     0      0    0     0    0    0     0  
3        0    0     0      0    0     0    0    0     0  
4        0    0     0      0    0     0    0    0     0  
...    ...  ...   ...    ...  ...   ...  ...  ...   ...  
5567     0    1     0      0    0     0    0    0     0  
5568     0    0     0      0    0     0    0    0     0  
5569     0    0     0      0    0     0    0    0     0  
5570     0    0     0      0    0     0    0    0     0  
5571     0    0     0      0    0     0    0    0     0  

[5572 rows x 214 columns]

In [89]:
# define df as X
X=df
X

150p  50  about  after  again  all  already  also  always  am  ...  win  \
0        0   0      0      0      0    0        0     0       0   0  ...    0   
1        0   0      0      0      0    0        0     0       0   0  ...    0   
2        0   0      0      0      0    0        0     0       0   0  ...    1   
3        0   0      0      0      0    0        1     0       0   0  ...    0   
4        0   0      0      0      0    0        0     0       0   0  ...    0   
...    ...  ..    ...    ...    ...  ...      ...   ...     ...  ..  ...  ...   
5567     0   0      0      0      0    0        0     0       0   0  ...    0   
5568     0   0      0      0      0    0        0     0       0   0  ...    0   
5569     0   0      0      0      0    0        0     0       0   0  ...    0   
5570     0   0      0      0      0    0        0     0       0   0  ...    0   
5571     0   0      0      0      0    0        0     0       0   0  ...    0   

      with  won  work  would  www  yeah  yes  you  your  
0        0    0     0      0    0     0    0    0     0  
1        0    0     0      0    0     0    0    0     0  
2        0    0     0      0    0     0    0    0     0  
3        0    0     0      0    0     0    0    0     0  
4        0    0     0      0    0     0    0    0     0  
...    ...  ...   ...    ...  ...   ...  ...  ...   ...  
5567     0    1     0      0    0     0    0    0     0  
5568     0    0     0      0    0     0    0    0     0  
5569     0    0     0      0    0     0    0    0     0  
5570     0    0     0      0    0     0    0    0     0  
5571     0    0     0      0    0     0    0    0     0  

[5572 rows x 214 columns]

In [90]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [91]:
# [1] with desision tree 
destree(X,y)

decision tree  0.9599282296650717
Execution time: 0.12089872360229492 seconds


In [92]:
# [2] with svc 
svc(X,y)

svc  0.9778708133971292
Execution time: 0.608966588973999 seconds


In [93]:
# [3] with mlp
mlp(X,y)

mlp 0.9772727272727273
Execution time: 2.101001739501953 seconds


In [94]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9473684210526315
Execution time: 0.22162437438964844 seconds


In [95]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9467703349282297
Execution time: 8.549813032150269 seconds


In [96]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [97]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9784688995215312
Execution time: 0.5058634281158447 seconds


In [98]:
# [8] with abc
abc(X,y)

AdaBoost 0.9730861244019139
Execution time: 0.445406436920166 seconds


In [99]:
# [9] with gaussianNB
gnb(X,y)

GaussianNB 0.6202153110047847
Execution time: 0.026091814041137695 seconds


In [100]:
# [10] with quardratic discriminant analysis
gda(X,y)

QuadraticDiscriminantAnalysis 0.6979665071770335
Execution time: 0.07932281494140625 seconds


C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


## CountVectorizer using max_feature 
* Try seven with change the lower case to false in countvectorizer 
     * worst one was [ GaussianNB ] with accuracy = 80.80%
     * best one was [ MLP ] with accuracy = 98.92%

In [101]:
count_vect = CountVectorizer(max_features=1000)
count_matrix = count_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data= count_array,columns=count_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  03  04  0800  08000839402  08000930705  10  100  1000  ...  \
0      0    0   0   0     0            0            0   0    0     0  ...   
1      0    0   0   0     0            0            0   0    0     0  ...   
2      0    0   0   0     0            0            0   0    0     0  ...   
3      0    0   0   0     0            0            0   0    0     0  ...   
4      0    0   0   0     0            0            0   0    0     0  ...   
...   ..  ...  ..  ..   ...          ...          ...  ..  ...   ...  ...   
5567   0    0   0   0     0            0            0   0    0     0  ...   
5568   0    0   0   0     0            0            0   0    0     0  ...   
5569   0    0   0   0     0            0            0   0    0     0  ...   
5570   0    0   0   0     0            0            0   0    0     0  ...   
5571   0    0   0   0     0            0            0   0    0     0  ...   

      yes  yesterday  yet  yo  you  your  yours  yourself  yr  yup  
0       0          0    0   0    0     0      0         0   0    0  
1       0          0    0   0    0     0      0         0   0    0  
2       0          0    0   0    0     0      0         0   0    0  
3       0          0    0   0    0     0      0         0   0    0  
4       0          0    0   0    0     0      0         0   0    0  
...   ...        ...  ...  ..  ...   ...    ...       ...  ..  ...  
5567    0          0    0   0    0     0      0         0   0    0  
5568    0          0    0   0    0     0      0         0   0    0  
5569    0          0    0   0    0     0      0         0   0    0  
5570    0          0    0   0    0     0      0         0   0    0  
5571    0          0    0   0    0     0      0         0   0    0  

[5572 rows x 1000 columns]

In [102]:
# define df as X
X=df
X

00  000  03  04  0800  08000839402  08000930705  10  100  1000  ...  \
0      0    0   0   0     0            0            0   0    0     0  ...   
1      0    0   0   0     0            0            0   0    0     0  ...   
2      0    0   0   0     0            0            0   0    0     0  ...   
3      0    0   0   0     0            0            0   0    0     0  ...   
4      0    0   0   0     0            0            0   0    0     0  ...   
...   ..  ...  ..  ..   ...          ...          ...  ..  ...   ...  ...   
5567   0    0   0   0     0            0            0   0    0     0  ...   
5568   0    0   0   0     0            0            0   0    0     0  ...   
5569   0    0   0   0     0            0            0   0    0     0  ...   
5570   0    0   0   0     0            0            0   0    0     0  ...   
5571   0    0   0   0     0            0            0   0    0     0  ...   

      yes  yesterday  yet  yo  you  your  yours  yourself  yr  yup  
0       0          0    0   0    0     0      0         0   0    0  
1       0          0    0   0    0     0      0         0   0    0  
2       0          0    0   0    0     0      0         0   0    0  
3       0          0    0   0    0     0      0         0   0    0  
4       0          0    0   0    0     0      0         0   0    0  
...   ...        ...  ...  ..  ...   ...    ...       ...  ..  ...  
5567    0          0    0   0    0     0      0         0   0    0  
5568    0          0    0   0    0     0      0         0   0    0  
5569    0          0    0   0    0     0      0         0   0    0  
5570    0          0    0   0    0     0      0         0   0    0  
5571    0          0    0   0    0     0      0         0   0    0  

[5572 rows x 1000 columns]

In [103]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [104]:
# [1] with desision tree 
destree(X,y)


decision tree  0.9700956937799043
Execution time: 0.6394410133361816 seconds


In [105]:
# [2] with svc 
svc(X,y)

svc  0.9832535885167464
Execution time: 2.4891207218170166 seconds


In [106]:
# [3] with mlp
mlp(X,y)

mlp 0.9892344497607656
Execution time: 2.807171106338501 seconds


In [107]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9240430622009569
Execution time: 0.29374194145202637 seconds


In [108]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9479665071770335
Execution time: 14.191742181777954 seconds


In [109]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [110]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9820574162679426
Execution time: 1.5079493522644043 seconds


In [111]:
# [8] with abc
abc(X,y)

AdaBoost 0.9700956937799043
Execution time: 2.290194511413574 seconds


In [112]:
# [9] with gaussianNB
gnb(X,y)

GaussianNB 0.8080143540669856
Execution time: 0.10300993919372559 seconds


In [113]:
# [10] with quardratic discriminant analysis
gda(X,y)

QuadraticDiscriminantAnalysis 0.9742822966507177
Execution time: 0.7039594650268555 seconds


C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


## TfidVectorizer : Feature Generation using TF-IDF
* Try eight with change the lower case to false in countvectorizer 
     * worst one was [ GaussianNB ] with accuracy = 80.80%
     * best one was [ MLP ] with accuracy = 98.92%

In [114]:
from sklearn.feature_extraction.text import TfidfVectorizer
# create the transform

vectorizer = TfidfVectorizer()
matrix =vectorizer.fit_transform(data.Message)
array = matrix.toarray()
new_data = pd.DataFrame(data=array,columns=vectorizer.get_feature_names())
new_data

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
1     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
3     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
...   ...  ...     ...           ...   ...   ...          ...          ...   
5567  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5568  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5569  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5570  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5571  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

      0125698789   02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
1            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
2            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
3            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
4            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
...          ...  ...  ...    ...     ...  ...        ...   ...   ...    ...   
5567         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5568         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5569         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5570         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5571         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   

       èn   ú1  〨ud  
0     0.0  0.0  0.0  
1     0.0  0.0  0.0  
2     0.0  0.0  0.0  
3     0.0  0.0  0.0  
4     0.0  0.0  0.0  
...   ...  ...  ...  
5567  0.0  0.0  0.0  
5568  0.0  0.0  0.0  
5569  0.0  0.0  0.0  
5570  0.0  0.0  0.0  
5571  0.0  0.0  0.0  

[5572 rows x 8709 columns]

In [115]:
# define df as X
X=df
X

00  000  03  04  0800  08000839402  08000930705  10  100  1000  ...  \
0      0    0   0   0     0            0            0   0    0     0  ...   
1      0    0   0   0     0            0            0   0    0     0  ...   
2      0    0   0   0     0            0            0   0    0     0  ...   
3      0    0   0   0     0            0            0   0    0     0  ...   
4      0    0   0   0     0            0            0   0    0     0  ...   
...   ..  ...  ..  ..   ...          ...          ...  ..  ...   ...  ...   
5567   0    0   0   0     0            0            0   0    0     0  ...   
5568   0    0   0   0     0            0            0   0    0     0  ...   
5569   0    0   0   0     0            0            0   0    0     0  ...   
5570   0    0   0   0     0            0            0   0    0     0  ...   
5571   0    0   0   0     0            0            0   0    0     0  ...   

      yes  yesterday  yet  yo  you  your  yours  yourself  yr  yup  
0       0          0    0   0    0     0      0         0   0    0  
1       0          0    0   0    0     0      0         0   0    0  
2       0          0    0   0    0     0      0         0   0    0  
3       0          0    0   0    0     0      0         0   0    0  
4       0          0    0   0    0     0      0         0   0    0  
...   ...        ...  ...  ..  ...   ...    ...       ...  ..  ...  
5567    0          0    0   0    0     0      0         0   0    0  
5568    0          0    0   0    0     0      0         0   0    0  
5569    0          0    0   0    0     0      0         0   0    0  
5570    0          0    0   0    0     0      0         0   0    0  
5571    0          0    0   0    0     0      0         0   0    0  

[5572 rows x 1000 columns]

In [116]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [117]:
# [1] with desision tree 
destree(X,y)

decision tree  0.9683014354066986
Execution time: 0.8521356582641602 seconds


In [118]:
# [2] with svc 
svc(X,y)

svc  0.9832535885167464
Execution time: 2.776510715484619 seconds


In [119]:
# [3] with mlp
mlp(X,y)

mlp 0.9892344497607656
Execution time: 2.8195695877075195 seconds


In [120]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9240430622009569
Execution time: 0.30895566940307617 seconds


In [121]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9479665071770335
Execution time: 14.362208127975464 seconds


In [122]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'


In [123]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9820574162679426
Execution time: 1.4717731475830078 seconds


In [124]:
# [8] with abc
abc(X,y)

AdaBoost 0.9706937799043063
Execution time: 2.1478872299194336 seconds


In [125]:
# [9] with gaussianNB
gnb(X,y)

GaussianNB 0.8080143540669856
Execution time: 0.08385992050170898 seconds


In [126]:
# [10] with quardratic discriminant analysis
gda(X,y)

QuadraticDiscriminantAnalysis 0.9742822966507177
Execution time: 0.629084587097168 seconds


C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


## here i will combine some method to gather   

# stop_word with max_df
* Try nine with change the lower case to false in countvectorizer 
     * worst one was [ QuadraticDiscriminantAnalysis] with accuracy = 48%
     * best one was [ MLP ] with accuracy = 98.92%

In [127]:
coun_vect = CountVectorizer(stop_words= 'english',max_df=0.65)
count_matrix = coun_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data=count_array,columns = coun_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8440 columns]

In [128]:
# define df as X
X=df
X

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8440 columns]

In [129]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [130]:
# [1] with desision tree 
destree(X,y)

decision tree  0.9688995215311005
Execution time: 14.981881141662598 seconds


In [131]:
# [2] with svc 
svc(X,y)

svc  0.979066985645933
Execution time: 30.338834524154663 seconds


In [132]:
# [3] with mlp
mlp(X,y)

mlp 0.9892344497607656
Execution time: 35.949175119400024 seconds


In [133]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9084928229665071
Execution time: 1.1751880645751953 seconds


In [134]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.9324162679425837
Execution time: 84.17634534835815 seconds


In [135]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [136]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9772727272727273
Execution time: 20.999879121780396 seconds


In [137]:
# [8] with abc
abc(X,y)

AdaBoost 0.9712918660287081
Execution time: 25.140809297561646 seconds


In [138]:
# [9] with gaussianNB
gnb(X,y)

GaussianNB 0.8953349282296651
Execution time: 0.6948840618133545 seconds


In [139]:
# [10] with quardratic discriminant analysis
gda(X,y)

C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


QuadraticDiscriminantAnalysis 0.48863636363636365
Execution time: 43.4179961681366 seconds


##  stop word with min_df 
* Try ten with change the lower case to false in countvectorizer 
     * worst one was [ GaussianNB] with accuracy = 53.16%
     * best one was [ MLP ] with accuracy = 98.92%


In [140]:
coun_vect = CountVectorizer(stop_words= 'english',min_df=0.01)
count_matrix = coun_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data=count_array,columns = coun_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


150p  50  amp  ask  babe  buy  care  cash  claim  com  ...  want  wat  \
0        0   0    0    0     0    0     0     0      0    0  ...     0    1   
1        0   0    0    0     0    0     0     0      0    0  ...     0    0   
2        0   0    0    0     0    0     0     0      0    0  ...     0    0   
3        0   0    0    0     0    0     0     0      0    0  ...     0    0   
4        0   0    0    0     0    0     0     0      0    0  ...     0    0   
...    ...  ..  ...  ...   ...  ...   ...   ...    ...  ...  ...   ...  ...   
5567     0   0    0    0     0    0     0     0      1    0  ...     0    0   
5568     0   0    0    0     0    0     0     0      0    0  ...     0    0   
5569     0   0    0    0     0    0     0     0      0    0  ...     0    0   
5570     0   0    0    0     0    0     0     0      0    0  ...     0    0   
5571     0   0    0    0     0    0     0     0      0    0  ...     0    0   

      way  week  win  won  work  www  yeah  yes  
0       0     0    0    0     0    0     0    0  
1       0     0    0    0     0    0     0    0  
2       0     0    1    0     0    0     0    0  
3       0     0    0    0     0    0     0    0  
4       0     0    0    0     0    0     0    0  
...   ...   ...  ...  ...   ...  ...   ...  ...  
5567    0     0    0    1     0    0     0    0  
5568    0     0    0    0     0    0     0    0  
5569    0     0    0    0     0    0     0    0  
5570    0     1    0    0     0    0     0    0  
5571    0     0    0    0     0    0     0    0  

[5572 rows x 106 columns]

In [141]:
# define df as X
X=df
X

150p  50  amp  ask  babe  buy  care  cash  claim  com  ...  want  wat  \
0        0   0    0    0     0    0     0     0      0    0  ...     0    1   
1        0   0    0    0     0    0     0     0      0    0  ...     0    0   
2        0   0    0    0     0    0     0     0      0    0  ...     0    0   
3        0   0    0    0     0    0     0     0      0    0  ...     0    0   
4        0   0    0    0     0    0     0     0      0    0  ...     0    0   
...    ...  ..  ...  ...   ...  ...   ...   ...    ...  ...  ...   ...  ...   
5567     0   0    0    0     0    0     0     0      1    0  ...     0    0   
5568     0   0    0    0     0    0     0     0      0    0  ...     0    0   
5569     0   0    0    0     0    0     0     0      0    0  ...     0    0   
5570     0   0    0    0     0    0     0     0      0    0  ...     0    0   
5571     0   0    0    0     0    0     0     0      0    0  ...     0    0   

      way  week  win  won  work  www  yeah  yes  
0       0     0    0    0     0    0     0    0  
1       0     0    0    0     0    0     0    0  
2       0     0    1    0     0    0     0    0  
3       0     0    0    0     0    0     0    0  
4       0     0    0    0     0    0     0    0  
...   ...   ...  ...  ...   ...  ...   ...  ...  
5567    0     0    0    1     0    0     0    0  
5568    0     0    0    0     0    0     0    0  
5569    0     0    0    0     0    0     0    0  
5570    0     1    0    0     0    0     0    0  
5571    0     0    0    0     0    0     0    0  

[5572 rows x 106 columns]

In [142]:
# define target column "Category" as y 
y=data['Category']
y

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: Category, Length: 5572, dtype: object

In [143]:
# [1] with desision tree 
destree(X,y)

decision tree  0.9617224880382775
Execution time: 0.061095237731933594 seconds


In [144]:
# [2] with svc 
svc(X,y)

svc  0.9706937799043063
Execution time: 0.34795379638671875 seconds


In [145]:
# [3] with mlp
mlp(X,y)

mlp 0.9736842105263158
Execution time: 2.493929386138916 seconds


C:\Users\emana\anaconda3\lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:702: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [146]:
# [4] with knighbors 
knb(X,y)

KNeighbors 0.9515550239234449
Execution time: 0.1657121181488037 seconds


In [147]:
# [5] with gaussian process 
gpc(X,y)

GaussianProcess 0.965311004784689
Execution time: 7.793372869491577 seconds


In [148]:
# [6] with RBF
#rbf(X,y)
#AttributeError: 'RBF' object has no attribute 'fit'

In [149]:
# [7] with random forest 
rfc(X,y)

RandomForest 0.9688995215311005
Execution time: 0.45908260345458984 seconds


In [150]:
# [8] with abc
abc(X,y)

AdaBoost 0.9659090909090909
Execution time: 0.27699923515319824 seconds


In [151]:
# [9] with gaussianNB
gnb(X,y)

GaussianNB 0.5316985645933014
Execution time: 0.017098665237426758 seconds


In [152]:
# [10] with quardratic discriminant analysis
gda(X,y)

QuadraticDiscriminantAnalysis 0.5394736842105263
Execution time: 0.03028702735900879 seconds


C:\Users\emana\anaconda3\lib\site-packages\sklearn\discriminant_analysis.py:887: UserWarning: Variables are collinear
  warnings.warn("Variables are collinear")


## here i will try one of the best result in loop

In [153]:
count_vect = CountVectorizer(stop_words='english')
count_matrix = count_vect.fit_transform(data.Message)
count_array = count_matrix.toarray()
df = pd.DataFrame(data= count_array,columns=count_vect.get_feature_names())
df

C:\Users\emana\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0      0    0       0             0     0     0            0            0   
1      0    0       0             0     0     0            0            0   
2      0    0       0             0     0     0            0            0   
3      0    0       0             0     0     0            0            0   
4      0    0       0             0     0     0            0            0   
...   ..  ...     ...           ...   ...   ...          ...          ...   
5567   0    0       0             0     0     0            0            0   
5568   0    0       0             0     0     0            0            0   
5569   0    0       0             0     0     0            0            0   
5570   0    0       0             0     0     0            0            0   
5571   0    0       0             0     0     0            0            0   

      0125698789  02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0              0   0  ...      0       0    0          0     0     0      0   
1              0   0  ...      0       0    0          0     0     0      0   
2              0   0  ...      0       0    0          0     0     0      0   
3              0   0  ...      0       0    0          0     0     0      0   
4              0   0  ...      0       0    0          0     0     0      0   
...          ...  ..  ...    ...     ...  ...        ...   ...   ...    ...   
5567           0   0  ...      0       0    0          0     0     0      0   
5568           0   0  ...      0       0    0          0     0     0      0   
5569           0   0  ...      0       0    0          0     0     0      0   
5570           0   0  ...      0       0    0          0     0     0      0   
5571           0   0  ...      0       0    0          0     0     0      0   

      èn  ú1  〨ud  
0      0   0    0  
1      0   0    0  
2      0   0    0  
3      0   0    0  
4      0   0    0  
...   ..  ..  ...  
5567   0   0    0  
5568   0   0    0  
5569   0   0    0  
5570   0   0    0  
5571   0   0    0  

[5572 rows x 8440 columns]

In [154]:
import numpy as np
X = df
y = data['Category']




acc_list = []
N = 20
st = time.time()
for i in range(N):
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

    MLP = MLPClassifier()
    MLP.fit(X_train, y_train)
    y_pred = MLP.predict(X_test)
    acc = accuracy_score(y_test, y_pred)
    acc_list.append(acc)
    print(acc)
    
et = time.time()
elapsed_time=et-st
np.mean(acc_list)
print('Execution time:', elapsed_time, 'seconds')

0.9850478468899522
0.9760765550239234
0.9820574162679426
0.9838516746411483
0.9814593301435407
0.9796650717703349
0.9820574162679426
0.9814593301435407
0.9850478468899522
0.9832535885167464
0.9808612440191388
0.9832535885167464
0.9838516746411483
0.9808612440191388
0.979066985645933
0.9826555023923444
0.9850478468899522
0.9772727272727273
0.986244019138756
0.9826555023923444
Execution time: 659.7179868221283 seconds
